Tests/FluentAssertions.Specs/Execution/TestFrameworkProviderTests.cs

```diff
@@ -0,0 +1,78 @@
+    {
+        var configurationStore = new TestConfigurationStore { Data = { [Configuration.TestFrameworkConfigurationKey] = "xunit2" } };
+        var configuration = new Configuration(configurationStore);
+        var result = TestFrameworkProvider.DetectFramework(configuration);
+
+        result.IsAvailable.Should().BeTrue();
+        result.Should().BeOfType<XUnit2TestFramework>();
+    }
+
+    [Fact]
+    public void When_running_test_with_unknown_test_framework_it_should_throw()
+    {
+        var configurationStore = new TestConfigurationStore { Data = { [Configuration.TestFrameworkConfigurationKey] = "foo" } };
+        var configuration = new Configuration(configurationStore);
+
+        this.Invoking(_ => TestFrameworkProvider.AttemptToDetectUsingAppSetting(configuration))
+            .Should().Throw<InvalidOperationException>()
+            .WithMessage("FluentAssertions was configured to use the test framework 'foo' but this is not supported."
+                         + " Please use one of the supported frameworks: mspec, nspec3, nunit, mstestv2, xunit2.");
+    }
+
+    [Fact]
+    public void When_running_test_with_direct_bound_but_unavailable_test_framework_it_should_throw()
+    {
+        var configurationStore = new TestConfigurationStore { Data = { [Configuration.TestFrameworkConfigurationKey] = "nspec3" } };
+        var configuration = new Configuration(configurationStore);
+
+        this.Invoking(_ => TestFrameworkProvider.AttemptToDetectUsingAppSetting(configuration))
+            .Should().Throw<InvalidOperationException>()
+            .WithMessage("FluentAssertions was configured to use the test framework 'nspec3' but it could not be found."
+                         + " Please use one of the supported frameworks: mspec, nspec3, nunit, mstestv2, xunit2.");
+    }
+
+    [Fact]
+    public void When_running_test_with_late_bound_but_unavailable_test_framework_it_should_throw()
+    {
+        var configurationStore = new TestConfigurationStore { Data = { [Configuration.TestFrameworkConfigurationKey] = "nunit" } };
+        var configuration = new Configuration(configurationStore);
+
+        this.Invoking(_ => TestFrameworkProvider.AttemptToDetectUsingAppSetting(configuration))
+            .Should().Throw<InvalidOperationException>()
+            .WithMessage("FluentAssertions was configured to use the test framework 'nunit' but the required assembly 'nunit.framework' could not be found."
+                         + " Please use one of the supported frameworks: mspec, nspec3, nunit, mstestv2, xunit2.");
+    }
+
+    private class TestConfigurationStore : IConfigurationStore
+    {
+        public IDictionary<string, string> Data { get; } = new Dictionary<string, string>();
+
+        public string GetSetting(string name)
+        {
+            Data.TryGetValue(name, out var setting);
+            return setting ?? string.Empty;
+        }
+    }
```

Comment:
To avoid relying on `Configuration.TestFrameworkConfigurationKey` and removes the need to extract it to a constant.
```cs
private sealed class TestConfigurationStore : IConfigurationStore
{
    private readonly string testFramework;

    public TestConfigurationStore(string testFramework)
    {
        this.testFramework = testFramework;
    }

    public string GetSetting(string name) => testFramework;
}
```

Reply:
What is the problem in defining the constant?
I think this is the better way at all instead if using literals, isn't it?

Reply:
The problem is tests using internals when not needed.
I had an extra look at this and we can use `Configuration.TestFrameworkName` instead.

```cs
[Fact]
public void When_running_xunit_test_explicitly_it_should_be_detected()
{
    var configuration = new Configuration(new TestConfigurationStore())
    {
        TestFrameworkName = "xunit2"
    };
    var result = TestFrameworkProvider.DetectFramework(configuration);

    result.IsAvailable.Should().BeTrue();
    result.Should().BeOfType<XUnit2TestFramework>();
}

private sealed class TestConfigurationStore : IConfigurationStore
{
    string IConfigurationStore.GetSetting(string name) => string.Empty;
}
```

Reply:
done